### PR TITLE
mui theme dark & 3rd party

### PR DESCRIPTION
**Pull request by Jackbennett**
Hi, I've updated the frontend to use the currently set theme on the ThemeProvider framework. This still needs work but 

Tasks involved:
- [x] Migrate `mender-theme.js` into a MUI light/dark
- [x] Remove `theme` access from outside `useTheme`
- [ ] Remove `colors` access from outside `useTheme` either `theme.palette` or custom `theme.colors`
- [x] Remove chartColorPalette for a mechanism from `useTheme` `theme.*`. Added `theme.palette.qualitative` but not in use yet
- [x] Somehow coordinate colors from `variables.less` into  design tokens used within the mui theme. 
* ~~Create interface to theme switching mechanism~~ not in scope
- [x] Format commits in line with contributors guide.

I made the changes from 3.1.0 but rebased onto master. I think I've kept the right changes to templates.

## uniting `*.less` and material theme

Resolution: Have wrapped CssBaseline in a styled component to add `:root` css variables as needed for entries from mui's theme. Less can use these variables. Less functions can have issues computing them but CSS functions should suffice.

Two ways I'm not a huge fan of might be;
* Remove all less in flavor of style components - Could be lots of work
* Remove all the colours from less, instead changing components to use material styles - probably difficult given the specificity in less

Initially made the changes to mapping material design tokens into less.
1. Add the tokens into variables.less,
2. Change existing variables to refer to those tokens
3. Update mui to pull the theme values from that less token map
4. Gradually over time change UI into styled components to use the mui tokens directly instead of less. Possibly after moving to bootstrap5 to avoid adding and jss

This issue with this is when switching to a 3rd parts mui the less does not reflect major changes of color.

## Concerns
* Want to avoid writing a bunch more styled components if that's all going to need changing again from Material v4 -> v5
* Light should look exactly the same as it does now. This isn't a rebrand.

## Related Discussions;
- Initial request; https://hub.mender.io/t/switchable-gui-material-theme/4471/4

# Previews
image
Current biggest issue with this being the greyscale used from less instead of `theme.palette.grey[x]`

image
Updated grey palette and token uses.

image
Using CSS variables for colors can switch custom secondary color PoC.

### CODE REVIEW

**Review by mzedel — CHANGES_REQUESTED** [review]
Thank you very much for the PR - that is quite a bit!

To continue on this I think targeting the feature branch would be best + considering the additional things to do it might be helpful to first do a _rough_ MUI upgrade and rebase this PR on that... What do you think @Jackbennett?

File: docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,43 @@
+volumes:
```
[the remark this answers is not in the document]
Comment (Jackbennett) [reply docker-compose.yml:1]: Ah yeh anything to do with docker I should drop in a rebase that's just my dev setup has snuck in there sorry.

**Comment by Jackbennett** [comment]
> To continue on this I think targeting the feature branch would be best + considering the additional things to do it might be helpful to first do a _rough_ MUI upgrade and rebase this PR on that... What do you think @Jackbennett?

I'm unsure as to how much churn you want into releases but happy to follow what make sense for you.

I'm focusing on linking up the context that mui and `less` effect so it should get easier and easier to push things into mui styled components. With regards to actual colors and styling no doubt that's up to yourselves internally what/if you want a dark theme to look like. I'm just getting the supports in place for themes in general.  You might choose to put the manager behind a plan addon, I won't write a manager I can edit the self-hosted service.

I have a question though, is all of the frontend code here? You don't have another enterprise source? Tracking down where some of the less is used there's some dead ends that could be removed.

Personally I haven't used material v5 yet only v4 but happy to get stuck in.

File: Dockerfile [hunk 2]
```diff
@@ -5,17 +5,18 @@ RUN npm ci
 COPY . ./
 RUN npm run build
 
-FROM nginx:1.21.5-alpine
+FROM nginx:1.21.5-alpine as Server
```
[the remark this answers is not in the document]
Comment (Jackbennett) [reply Dockerfile]: I toyed around with the entrypoint but didn't seem to need it. Not for dev, that env.js being written in there makes things a bit complicated. Docker stuff is outside the scope anyhow so I've rebased it out.

File: src/js/components/devices/device-details/notifications.js [hunk 3]
```diff
@@ -19,45 +18,52 @@ const monitoringSeverities = {
 };
 
 export const severityMap = {
-  [monitoringSeverities.CRITICAL]: { className: 'red', icon: errorIcon, listIcon: <ErrorIcon className="red" /> },
-  [monitoringSeverities.CRITICAL_FLAPPING]: { className: '', icon: errorIcon, listIcon: <ErrorIcon className="red" /> },
-  [monitoringSeverities.OK]: { className: '', icon: successIcon, listIcon: <CheckIcon className="green" /> },
-  [monitoringSeverities.UNKNOWN]: { className: '', icon: questionIcon, listIcon: <HelpIcon /> }
+  [monitoringSeverities.CRITICAL]: { className: 'red', icon: errorIcon },
```
[the remark this answers is not in the document]
Comment (Jackbennett) [reply src/js/components/devices/device-details/notifications.js:21]: Fixed, spoiled by typescript didn't even read that was an export sorry.

**Comment by Jackbennett** [comment]
This is possibly in a good spot to update to mui v5 to start stripping colours from main.less and stop using `import { colors }`

mui can get the theme from `<Box sx={background-color: theme => theme.palette.grey[200]}} />` which would probably remove 80% of the `import { useTheme } ... const theme = useTheme()` I just added.

I don't think I'm going to pick this up for another week which sounds like nice timing while you had this scheduled. I'm not sure if a small update first would make this easier, other than getting the component interfaces out there. It would introduce a short time to notice regressions from the UI that hasn't changed in a year (guessing from commit dates i saw while editing) a 3.x with small changes and 3.x+1 on mui-v5

File: src/js/components/artifacts/releaseslist.js [hunk 4]
```diff
@@ -27,7 +27,7 @@ const ReleaseListItem = ({ data, index, style }) => {
   let isSelected = index === 0;
   isSelected = release && selectedRelease ? release.Name === selectedRelease.Name : isSelected;
   return (
-    <ListItem button className={`repository-list-item ${isSelected ? 'active' : ''}`} onClick={() => onSelect(release)} style={style}>
+    <ListItem button selected={isSelected} className={['repository-list-item']} onClick={() => onSelect(release)} style={style}>
```
Comment (mzedel) [inline src/js/components/artifacts/releaseslist.js:30]: that should most likely be a simple string

File: src/js/components/deployments/deployment-wizard/softwaredevices.js [hunk 5]
```diff
@@ -24,10 +23,13 @@ export const styles = {
...
     marginBottom: 0
   },
+}));
+
+const hardCodedStyle = {
   textField: {
     minWidth: 400
   }
-};
+}
```
Comment (mzedel) [inline src/js/components/deployments/deployment-wizard/softwaredevices.js:32]: that omission will be flagged by the prettier checks

File: src/js/components/deployments/deployment-wizard/softwaredevices.js [hunk 6]
```diff
@@ -24,10 +23,13 @@ export const styles = {
   selectionTitle: {
     marginBottom: 0
   },
```
Comment (mzedel) [inline src/js/components/deployments/deployment-wizard/softwaredevices.js:25]: that comma will be flagged by the prettier checks

**Review by mzedel — CHANGES_REQUESTED** [review]
Thanks once again for the continued changes! I progressed a bit with the upgrade to mui5, however the changes to the components (alone the imports e.g.), test files and theme structure are a lot...
In order to streamline the adjustments I would thus like to merge your PR in the feature branch, get the rough mui upgrade in there (after I rebased again with your changes merged) as well and then from there get rid of the remaining inconsistencies in the theme... would that work for you?

**Comment by Jackbennett** [comment]
I had some fixes still on here to push in for what you mentioned.

I also found to write the theme into css variables from mui for less to pick back up as an interim to getting it rebuilt. I'm swapping between light/dark/custom to find elements that are missing when considering styling.

There's a lot that will change with the v5 and updating tests, I think what you propose sounds like the best way to handle that. To take advantage of v5 there's component changes and some stylistic choices I don't think I can make as an outsider. That pink in `tooltipText` I'm not sure about for example.

Sorry about the lint changes creeping in I've got the code in  a container but not VSCode so from its point of view it's not installed.

**Comment by mzedel** [comment]
Nice solution with the css variables, based on their availability and the general reduction of less dependency in the codebase that might be a good way to _some day_ get rid of it entirely (and speed up the build setup)...
Regarding the theme I will also gladly keep this for a colleague to tune as I tend to get lost in the code to the detriment of the design 🙈.
And don't worry about the lint changes - with the mui upgrade applied, there are worse tests to fix (besides the pending theme alignment).

If you're ok with where things stand feel free to mark this as ready for review, I'll merge to the feature branch and ready the  PR for the upgrade (the rebase is probably going to be interesting)...

**Comment by Jackbennett** [comment]
Thanks for your collaboration it helped clear a lot up :)

Some notes;

* You might not want to keep `theme.palette.qualitative` for chart colors
* Seems like CSS variables don't go into less functions life `fade` but there's alternates. Also consider just using the `.light` `.dark` variants along side `.main` that mui provides
* `import { colors } ...` is the next major thing to get rid of
* I think MUIv5 using function in `sx=` could remove most useTheme calls.
* I used this to harmonise the greyscale https://vis4.net/palettes/#/11|d|969696,aaaaaa,c7c7c7,f7f7f7|eaf4f3,d4e9e7|1|1 then took out one of the lighter stops that was low contrast

Hope it all helped.

**Comment by mzedel** [comment]
Alright, I'll keep that list prominent and will forward the palettes tip! I'll try to make this land in master asap and will ping here once there is a docker image pushed that you could use - if you don't use a self built one anyway of course.

There is however one last thing that you need to do (for legal reasons for us): adding a signoff to all your commits (& preferably also a last rebase onto the target branch - but at least that I could take care of if you don't want to)...

**Comment by mzedel** [comment]
Great, that last push fixed all commits except for the last one (dd645e5c6000e66d9d860fb6ccae2ecddb0c6609), where `Author: Jackben <j.bennett@engineeredarts.co.uk>` and `Signed-off-by: Jackbennett <code@jackben.net>` conflict - that `test:check-commits` stage in the pipeline is the only that needs to pass for now...

**Comment by Jackbennett** [comment]
Ok CI is happy with eslint fixed and commits with signoff and changelog I'll leave this alone now :+1:

**Review by mzedel — APPROVED** [review]

**Pull request merged**